File: linked_list/142.linked-list-cycle-ii.cpp
/*
 * @lc app=leetcode id=142 lang=cpp
 *
 * [142] Linked List Cycle II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
 public:
  ListNode *detectCycle(ListNode *head) {
    if (!head) return nullptr;

    ListNode *slow = head, *fast = head->next;
    while (fast and fast->next) {
      if (slow == fast) break;
      slow = slow->next;
      fast = fast->next->next;
    }

    if (slow != fast) return nullptr;
    slow = head;
    while (slow != fast->next) {
      slow = slow->next;
      fast = fast->next;
    }

    return slow;
  }
};
// @lc code=end

File: 142.linked-list-cycle-ii.py
#
# @lc app=leetcode id=142 lang=python3
#
# [142] Linked List Cycle II
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        slow, fast = head, head
        while fast and fast.next:
            if slow == fast.next:
                break
            slow = slow.next
            fast = fast.next.next
        else:
            return None
        
        fast = head
        while slow.next != fast:
            slow = slow.next
            fast = fast.next
        
        return fast



# @lc code=end

